swiftsimio/objects.py

```diff
@@ -26,6 +121,243 @@ def wrapped(self, *args, **kwargs):
+    elif ca1 and not ca2:
+        # two is not a cosmo_array, allow e.g. multiplication by constants:
+        return cf1
+    elif (ca1 and ca2) and ((cf1 is None) or (cf2 is None)):
+        # both cosmo_array but not both with cosmo_factor
+        # (both without shortcircuited above already):
+        warnings.warn(
+            f"Mixing ufunc arguments with and without cosmo_factors ({cf1} and {cf2}),"
+            f" discarding cosmo_factor in return value.",
+            RuntimeWarning,
+        )
+        return None
+    elif (ca1 and ca2) and ((cf1 is not None) and (cf2 is not None)):
+        # both cosmo_array and both with cosmo_factor:
+        return cf1 * cf2  # cosmo_factor.__mul__ raises if scale factors differ
+    else:
+        raise RuntimeError("Unexpected state, please report this error on github.")
+
+
+def _preserve_cosmo_factor(ca_cf1, ca_cf2=None, **kwargs):
+    ca1, cf1 = ca_cf1
+    ca2, cf2 = ca_cf2 if ca_cf2 is not None else (None, None)
+    if ca_cf2 is None:
+        # single argument, return promptly
+        return cf1
+    elif (cf1 is None) and (cf2 is None):
+        # neither has cosmo_factor information:
+        return None
+    elif ca1 and not ca2:
+        # only one is cosmo_array
+        return cf1
+    elif ca2 and not ca1:
+        # only one is cosmo_array
+        return cf2
+    elif (ca1 and ca2) and (cf1 is None and cf2 is not None):
+        # both cosmo_array, but not both with cosmo_factor
+        # (both without shortcircuited above already):
+        warnings.warn(
+            f"Mixing ufunc arguments with and without cosmo_factors, continuing assuming"
+            f" provided cosmo_factor ({cf2}) for all arguments.",
+            RuntimeWarning,
+        )
+        return cf2
+    elif (ca1 and ca2) and (cf1 is not None and cf2 is None):
+        # both cosmo_array, but not both with cosmo_factor
+        # (both without shortcircuited above already):
+        warnings.warn(
+            f"Mixing ufunc arguments with and without cosmo_factors, continuing assuming"
+            f" provided cosmo_factor ({cf1}) for all arguments.",
+            RuntimeWarning,
+        )
+        return cf1
+    elif (ca1 and ca2) and (cf1 != cf2):
+        raise ValueError(
+            f"Ufunc arguments have cosmo_factors that differ: {cf1} and {cf2}."
+        )
+    elif (ca1 and ca2) and (cf1 == cf2):
+        return cf1  # or cf2, they're equal
+    else:
+        raise RuntimeError("Unexpected state, please report this error on github.")
+
+
+def _power_cosmo_factor(ca_cf1, ca_cf2, inputs=None, power=None):
+    if inputs is not None and power is not None:
+        raise ValueError
+    ca1, cf1 = ca_cf1
+    ca2, cf2 = ca_cf2
+    power = inputs[1] if inputs else power
+    if hasattr(power, "units"):
+        if not power.units.is_dimensionless:
+            raise ValueError("Exponent must be dimensionless.")
+        else:
+            power.to_value(unyt.dimensionless)
```

Comment:
`to_value` returns a copy of the array converted to that unit. It does not affect the array itself... Perhaps `convert_to_value` is what you were looking for? Though this does mutate the external array's state which may be bad behavior. Can you just leave it in its original (dimensionless compatible) units?

Reply:
I think we're stuck with a copy operation here. This is determining what exponent to raise the units to. `convert_to_value` is a bad idea because it can unexpectedly modify the user's input data. Taking the raw value risks something like `power = 2000 * u.mm / (1 * u.m)`, when that gets used to exponentiate the units you'll get an exponent of `2000` instead of `2`. That leaves copying. I don't think it's a big deal because the thing being copied is the exponent argument of `np.power`, which in most cases is a scalar, and already dimensionless. Only if you had something like `np.power(2, np.arange(100000) * u.m / u.cm)` would you get an expensive copy, and it's hard to think of a use-case for such an expression. I guess I could add an `elif power.units is u.dimensionless` to avoid the copy if the units are explicitly the dimensionless unit?

Reply:
Well the problem here is that in the `else` case you're not actually using the converted value... You can do:

`use_power = power.to_value(unyt.dimensionless)`

but as the conversion is not done in-place right now this line does nothing. I think adding the elif for the simple case is a good idea.

Reply:
Well the problem here is that in the `else` case you're not actually using the converted value... You can do:

`use_power = power.to_value(unyt.dimensionless)`

but as the conversion is not done in-place right now this line does nothing. I think adding the elif for the simple case is a good idea.

Reply:
Um, yes, you're right of course. Sorry I missed the point on that one before, fixed now.